source/mir/random/engine/package.d

```diff
@@ -199,6 +270,147 @@ version(mir_random_test) unittest
     static assert(is(EngineReturnType!Random == size_t));
 }
 
+static if (THREAD_LOCAL_STORAGE_AVAILABLE)
+{
+    /++
+    Thread-local instance of the default $(LREF Random) allocated and seeded independently
+    for each thread. Requires $(LINK2 https://en.wikipedia.org/wiki/Thread-local_storage, TLS).
+    +/
+    alias rndGen = threadLocal!Random;
+    ///
+    @nogc nothrow @safe version(mir_random_test) unittest
+    {
+        import mir.random;
+
+        int[10] array;
+        foreach (ref e; array)
+            e = rndGen.rand!int;
+        auto picked = array[rndGen.randIndex(array.length)];
+
+        import std.complex;
+        auto c = complex(rndGen.rand!real, rndGen.rand!real);
+    }
+
+    /++
+    Thread-local instance of the specified random number generator allocated and seeded uniquely
+    for each thread. Requires $(LINK2 https://en.wikipedia.org/wiki/Thread-local_storage, TLS).
+
+    `_rawThreadLocal!Engine` is the potentially-uninitialized area of thread-local storage
+    used by `threadLocal!Engine`. Normally you will not need to use `_rawThreadLocal`:
+    if you access the generator this way it may not have been seeded yet. But if you need
+    to take the address of the generator, `_rawThreadLocal` is public so the compiler can
+    tell it is `@safe`.
+    +/
+    @property ref Engine threadLocal(Engine)()
+        if (isSaturatedRandomEngine!Engine && is(Engine == struct))
+    {
+        pragma(inline, true);
+        static bool initialized;
```

Comment:
This variable should be public to do not reproduce the phobos issue. Fibers (green threads) can be interrupted.
So a user code, including mir random functions like `randIndex`, should set `initialized` to false before they start using rng and set it to true after they finish. This will force reinitilization if a thread was terminated during engine state update.

Reply:
Hm, may be we can do it with a wrapper that hold TLS reference.

Reply:
With this change we can promote mir-random as better choice for vibe-d

Reply:
>This variable should be public to do not reproduce the phobos issue. Fibers (green threads) can be interrupted.

Is there a place I can read more about this issue?

>So a user code, including mir random functions like randIndex, should set initialized to false before they start using rng and set it to true after they finish. This will force reinitilization if a thread was terminated during engine state update.

Maybe your "TLS wrapper" idea could make this convenient: the wrapper's constructor (after ensuring the rng is set up) could set `initialized` to false, and the destructor could set `initialized` back to true.